private/tools/java/com/github/bazelbuild/rules_jvm_external/maven/MavenPublisher.java

```diff
@@ -261,6 +370,102 @@ private static String toHexS(String fmt, String algorithm, byte[] toHash) {
     }
   }
 
+  /**
+   * Attempts to download the file at the given targetUrl. Valid protocols are: http(s), file, and
+   * s3 at the moment.
+   */
+  private static CompletableFuture<Optional<String>> download(
+      String targetUrl, Credentials credentials) {
+    if (targetUrl.startsWith("http")) {
+      return httpDownload(targetUrl, credentials);
+    } else if (targetUrl.startsWith("file://")) {
+      return fileDownload(targetUrl);
+    } else if (targetUrl.startsWith("s3://")) {
+      return s3Download(targetUrl);
+    } else {
+      throw new IllegalArgumentException("Unsupported protocol for download.");
+    }
+  }
+
+  private static CompletableFuture<Optional<String>> s3Download(String targetUrl) {
+    return CompletableFuture.supplyAsync(
+        () -> {
+          S3Client s3Client = S3Client.create();
+          try {
+            URI s3Uri = new URI(targetUrl);
+            String bucketName = s3Uri.getHost();
+            String key = s3Uri.getPath().substring(1);
+            GetObjectRequest request =
+                GetObjectRequest.builder().bucket(bucketName).key(key).build();
+            ResponseInputStream<GetObjectResponse> s3Object = s3Client.getObject(request);
+            return Optional.of(
+                CharStreams.toString(new InputStreamReader(s3Object, StandardCharsets.UTF_8)));
+          } catch (IOException e) {
+            throw new UncheckedIOException(e);
+          } catch (URISyntaxException e) {
+            throw new RuntimeException(e);
+          } catch (S3Exception e) {
+            if (e.statusCode() == 404) {
+              return Optional.empty();
+            } else {
+              throw new RuntimeException(e);
+            }
+          }
+        });
+  }
+
+  private static CompletableFuture<Optional<String>> fileDownload(String targetUrl) {
+    return CompletableFuture.supplyAsync(
+        () -> {
+          try {
+            Path path = Paths.get(new URL(targetUrl).toURI());
+            if (!Files.exists(path)) {
+              return Optional.empty();
+            }
+            return Optional.of(Files.readString(path, StandardCharsets.UTF_8));
+          } catch (IOException e) {
+            throw new UncheckedIOException(e);
+          } catch (URISyntaxException e) {
+            throw new RuntimeException(e);
+          }
+        });
+  }
+
+  private static CompletableFuture<Optional<String>> httpDownload(
+      String targetUrl, Credentials credentials) {
+    return CompletableFuture.supplyAsync(
+        () -> {
+          try {
+            HttpTransport httpTransport = new NetHttpTransport();
```

Comment:
All versions of the JDK that we support have the native HTTP handler. We already use this in the `HttpDownloader` class, which also handles things like using the `.netrc` file if there is one. Perhaps use that? We could do that in a follow-up PR, but what I don't want to happen is for us to have a proliferation of http clients in this repo.